### ISSUE TITLE
Evaluate equivalent resources within the same source

### ISSUE THREAD

**Issue by kshefchek** [monarch-initiative/dipper#505]
There are cases where we say two IRIs from the same source are the same.  This is sometimes accurate but can lead to cycles in scigraph after clique merging.  Consider:

ZFIN:ZDB-FISH-150901-25626 owl:sameAs ZFIN:ZDB-GENO-041129-4
ZFIN:ZDB-FISH-150901-25626 GENO:0000382 ZFIN:ZDB-GENO-041129-4

GENO_0000382 ! has_variant_part

This results in a cycle in scigraph after the clique merge.  Theres also some odd equivalent logic occurring for this node where there are 5 has_variant_part edges pointing to itself.  This results in 325 different paths to just join this node to itself via BFO:0000051!*

    MATCH path=(genotype:Node{iri:'http://zfin.org/ZDB-FISH-150901-25626'})-[relation:`http://purl.obolibrary.org/obo/GENO_0000382`*]->(genotype:Node{iri:'http://zfin.org/ZDB-FISH-150901-25626'}) RETURN COUNT(DISTINCT(path))

I propose we remove owl:sameAs to connect fish to genotypes.  Even if these are technically the same, ZFIN has created separate resources (IRIs) for these and I'm not sure we can always guarantee equality.

For MGI, we have the scigraph logs:

    SEVERE: More than one node in a clique designated as leader. Using failover strategy to elect a leader.
    Aug 11, 2017 3:50:38 PM io.scigraph.owlapi.postprocessors.Clique electCliqueLeader
    SEVERE: http://www.informatics.jax.org/accession/MGI:5452661
    Aug 11, 2017 3:50:38 PM io.scigraph.owlapi.postprocessors.Clique electCliqueLeader
    SEVERE: http://www.informatics.jax.org/accession/MGI:3651550
    Aug 11, 2017 3:59:42 PM io.scigraph.owlapi.postprocessors.Clique electCliqueLeader
    SEVERE: More than one node in a clique designated as leader. Using failover strategy to elect a leader.
    Aug 11, 2017 3:59:42 PM io.scigraph.owlapi.postprocessors.Clique electCliqueLeader
    SEVERE: http://www.informatics.jax.org/accession/MGI:3648317
    Aug 11, 2017 3:59:42 PM io.scigraph.owlapi.postprocessors.Clique electCliqueLeader
    SEVERE: http://www.informatics.jax.org/accession/MGI:96741
     ...

cc @mbrush

**Comment by kshefchek** [comment]
I did a test where we removed the sameAs relation from ZFIN fish to genotypes and reloaded scigraph, here are the cypher query timings before (timed out indicates >5 hours):

gene-phenotype.yaml - 49.11 min - 4579044
pathway-phenotype.yaml - 3.546 h - 4663692
model-gene.yaml timed out!
variant-genotype.yaml timed out!
literature-gene.yaml timed out!
variant-phenotype.yaml timed out!
model-variant.yaml timed out!
literature-variant.yaml timed out!
genotype-gene.yaml timed out!

and after:

model-gene.yaml - 11.62 min - 262278
model-variant.yaml - 11.76 min - 419240
variant-genotype.yaml - 20.47 min - 750740
genotype-gene.yaml - 20.39 min - 422853
variant-phenotype.yaml - 35.71 min - 2969048
literature-variant.yaml - 35.71 min - 8631924
gene-gene.yaml - 39.03 min - 79312
pathway-phenotype.yaml - 40.32 min - 4288516
gene-phenotype.yaml - 45.32 min - 3004503

So safe to say the issue was these has_part cycles.  Although it is odd since we were making these eq statements in previous loads without issue.  There must be some errant sameAs relations along with the correct ones.  Regardless, even though part_of is reflexive, cycles will slow us down when generating paths, especially with one to many cardinality. 

For MGI, some of these look correct while others like the above example are wrong.  I'll take a closer look.

**Comment by mbrush** [comment]
Regarding the relationships we want to assert between ZDB-FISH and ZDB-GENO ids:
- in cases there the ZDB-FISH **does not** include extrinsic component (i.e. no morpholino treatment), the relation between the FISH and GENO is **owl:sameAs**.   
- in cases there the ZDB-FISH **does** include an extrinsic component, the relation between the FISH and GENO is **GENO:0000382** (has_variant_part).
 
Essentially, the owl:sameAs link between a ZDB-FISH id and a ZDB-GENO id should only be created when the FISH includes no extrinsic component - because in such cases  the ZDB-FISH id represents exactly the same thing as the ZDB-GENO id (e.g. https://zfin.org/ZDB-FISH-150901-14833).
 
This would mean that the scenario Kent laid out above shoudl never happen  (i.e. where there is both a owl:sameAs and GENO:0000382 relation between the same FISH and GENO IRI).  It seems however that the owl:sameAs assertion is being made every time - leading to many incorrect and performance-degrading owl:sameAs assertions between FISH and GENO ids. 
 
A fix where **all** owl:sameAs links are removed will result in phenotypes not being linked to FISH IDs where a link should exist, and/or phenotype not propagating from genotypes to genes (depending on whether the G2P association is asserted on the FISH or the GENO).  Perhaps a more precise fix would require just a simple rule that said:
 
`For each asserted FISH - GENO pair, create an owl:sameAs link if the FISH does not have_variant_part some GENO:0000524 ('extrinsic genotype') or  GENO:0000504 ('reagent targeted gene'), and create a GENO:0000382 link if it does.`
 
This should drastically reduce the number of sameAs assertions - and improve the clique-merge/load performance, but maintain important links in the data.
   
@kshefchck does this make sense to do?The we could write some unit test to confirm that there are no FISH-GENO pairs with both owl:sameAs and GENO:0000382 links between them.

**Comment by kshefchek** [comment]
I would argue for simplicity, ZDB-FISH has_variant_part ZDB-GENO.  

> A fix where all owl:sameAs links are removed will result in phenotypes not being linked to FISH IDs where a link should exist

These should be captured as model-phenotype relations?  We should add a pattern.

**Comment by kshefchek** [comment]
Here is a better outline of the issue:

Individuals
ZFIN:FISH123 ! some fish
ZFIN:GENO456 ! some genotype
bnode:NCBITaxon7955-with-ZFIN-FISH123 ! some model

Classes:
ZP:0004432 ! abnormal(ly) morphology vasculature
NCBITaxon:7955 ! Danio rerio
GENO:0000525 ! effective genotype
GENO:0000000 ! intrinsic genotype
GENO:0000611! genomic background


Edges
RO:0002162 ! in taxon
GENO:0000385 ! has reference part
GENO:0000382 ! has variant part
GENO:0000222 ! has genotype
RO:0002200 ! has phenotype
owl:sameAs ! sameAs

Current Model:

    bnode:NCBITaxon7955-with-ZFIN-FISH123 'has genotype' ZFIN:FISH123
    ZFIN:FISH123 in taxon 'Danio rerio'
    bnode:NCBITaxon7955-with-ZFIN-FISH123 a 'Danio rerio'

    ZFIN:FISH123 'has phenotype' 'abnormal(ly) morphology vasculature'

    if ZFIN:GENO456 is a wildtype:
        ZFIN:FISH123 'has reference part' ZFIN:GENO456
        ZFIN:GENO456 a 'genomic background'

    else:
       ZFIN:FISH123 'has variant part' ZFIN:GENO456
       ZFIN:GENO456 a 'intrinsic genotype'


    if extrinsic genoytype exists:
       ZFIN:FISH123 sameAs ZFIN:GENO456
    else:
       ZFIN:ZDB-FISH-161201-9 a 'effective genotype'


Option 1:
Remove any has_part relation if no extrinsic genotype exists (see @mbrush's comment above)

Option 2:
Fish ID becomes the organism (model), Geno ID is either a effective/intrinsic/background genotype

Option 3:
Fish ID is always the effective genotype, Geno ID is either a intrinsic or background genotype

My vote would be for option 2 or 3.  We can replace a bnode with a real ID, and get rid of some if/else logic in the code. In this case the app would show phenotype links to both the FISH and GENO ids, and model_of relations to the FISH id.

**Comment by kshefchek** [comment]
Looked in the two bugs above, the ZFIN issue is caused by ZDB-FISH-150901-25626 and ZDB-FISH-150901-26522 having identical fields in the fish components file (other than label):

Fish ID                |  Fish Name            |  Gene ID  |  Gene Symbol  |  Affector ID       |  Affector Symbol  |  Construct ID               |  Construct Symbol  |  Background ID  |  Background Name  |  Genotype ID        |  Genotype Name
-----------------------|-----------------------|-----------|---------------|--------------------|-------------------|-----------------------------|--------------------|-----------------|-------------------|---------------------|---------------
ZDB-FISH-150901-25626  |  vu12Tg + MO1-mir219  |           |               |  ZDB-ALT-041129-8  |  vu12Tg           |  ZDB-TGCONSTRCT-070117-167  |  Tg(olig2:EGFP)    |                 |                   |  ZDB-GENO-041129-4  |  vu12Tg
ZDB-FISH-150901-26522  |  vu12Tg               |           |               |  ZDB-ALT-041129-8  |  vu12Tg           |  ZDB-TGCONSTRCT-070117-167  |  Tg(olig2:EGFP)    |                 |                   |  ZDB-GENO-041129-4  |  vu12Tg


In the current code this results in:
ZDB-FISH-150901-25626 owl:sameAs ZDB-GENO-041129-4
ZDB-FISH-150901-26522 owl:sameAs ZDB-GENO-041129-4


For MGI, the issue comes from combining xrefs from MGI and NCBIGene, we get:

From MGI (this is correct):
MGI:3651550 owl:equivalentClass ENSEMBL:ENSMUSG00000087342
MGI:5452661 owl:equivalentClass ENSEMBL:ENSMUSG00000065061

From NCBI (the second triple is incorrect):
MGI:3651550 owl:equivalentClass ENSEMBL:ENSMUSG00000087342
MGI:3651550 owl:equivalentClass  ENSEMBL:ENSMUSG00000065061


For Mouse it is probably better for us to pull xrefs from MGI and ignore them from NCBI.  Both issues should be fixed in the next release.
